packages/flutter/lib/src/widgets/scroll_notification_observer.dart

```diff
@@ -0,0 +1,174 @@
+  void addListener(ScrollNotificationCallback listener) {
+    assert(_debugAssertNotDisposed());
+    _listeners!.add(_ListenerEntry(listener));
+  }
+
+  /// Remove the specified [ScrollNotificationCallback].
+  void removeListener(ScrollNotificationCallback listener) {
+    assert(_debugAssertNotDisposed());
+    for (final _ListenerEntry entry in _listeners!) {
+      if (entry.listener == listener) {
+        entry.unlink();
+        return;
+      }
+    }
+  }
+
+  void _notifyListeners(ScrollNotification notification) {
+    assert(_debugAssertNotDisposed());
+    if (_listeners!.isEmpty)
+      return;
+
+    final List<_ListenerEntry> localListeners = List<_ListenerEntry>.from(_listeners!);
+    for (final _ListenerEntry entry in localListeners) {
+      try {
+        if (entry.list != null)
+          entry.listener(notification);
+      } catch (exception, stack) {
+        FlutterError.reportError(FlutterErrorDetails(
+          exception: exception,
+          stack: stack,
+          library: 'widget library',
+          context: ErrorDescription('while dispatching notifications for $runtimeType'),
+          informationCollector: () sync* {
+            yield DiagnosticsProperty<ScrollNotificationObserverState>(
+              'The $runtimeType sending notification was',
+              this,
+              style: DiagnosticsTreeStyle.errorProperty,
+            );
+          },
+        ));
+      }
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return NotificationListener<ScrollNotification>(
+      onNotification: (ScrollNotification notification) {
+        _notifyListeners(notification);
+        return false;
+      },
+      child: _ScrollNotificationObserverScope(
+        scrollNotificationObserverState: this,
+        child: widget.child,
+      ),
+    );
+  }
+
+  @override
+  void dispose() {
+    assert(_debugAssertNotDisposed());
+    _listeners = null;
+    super.dispose();
+  }
+}
+
+/// Notifies its listeners when a descendant scrolls.
+///
+/// To add a listener to a [ScrollNotificationObserver] ancestor:
+/// ```dart
+/// void listener(ScrollNotification notification) {
+///   // Do something, maybe setState()
+/// }
+/// ScrollNotificationObserver.of(context).addListener(listener)
+/// ```
+///
+/// To remove the listener from a [ScrollNotificationObserver] ancestor:
+/// ```dart
+/// ScrollNotificationObserver.of(context).removeListener(listener);
+///```
+///
+/// Stateful widgets that share an ancestor [ScrollNotificationObserver] typically
+/// add a listener in [State.didChangeDependencies] (removing the old one
+/// if necessary) and remove the listener in their [State.dispose] method.
+///
+/// This widget is similar to [NotificationListener]. It supports
+/// a listener list instead of just a single listener and its listeners
+/// run unconditionally, they do not require a gating boolean return value.
+class ScrollNotificationObserver extends StatefulWidget {
```

Comment:
Should this new class have tests?

Reply:
Absolutely. I hadn't quite finished with this PR ...